homeassistant/components/samsungtv_upnp/media_player.py

```diff
@@ -0,0 +1,423 @@
+
+    @property
+    def source_list(self):
+        """List of available input sources."""
+        return list(self._source_list)
+
+    @property
+    def source(self):
+        """Name of the current input source."""
+        return self._source
+
+    @property
+    def media_content_type(self):
+        """Content type of current playing media."""
+        return MEDIA_TYPE_CHANNEL if self._media_channel else None
+
+    @property
+    def media_duration(self):
+        """Duration of current playing media in seconds."""
+        return self._media_duration
+
+    @property
+    def media_position(self):
+        """Position of current playing media in seconds."""
+        return self._media_position
+
+    @property
+    def media_position_updated_at(self):
+        """When was the position of the current playing media valid."""
+        return self._media_position_updated
+
+    @property
+    def media_channel(self):
+        """Channel currently playing"""
+        return self._media_channel
+
+    @property
+    def media_title(self):
+        """Title of current playing media."""
+        return self._media_title
+
+    @property
+    def media_content_type(self):
+        """Content type of current playing media."""
+        return MEDIA_TYPE_CHANNEL if self._media_channel else None
+
+    async def async_select_source(self, source):
+        """Select input source."""
+        if source not in self._source_list:
+            _LOGGER.debug('Unsupported source')
+            return
+        id = self._source_list[source]
+
+        action = self._device._action('MTVA', 'SetMainTVSource')
+        if not action:
+            _LOGGER.debug('Missing action MTVA/SetMainTVSource')
+            return
+
+        try:
+            result = await action.async_call(Source=source, ID=id, UiID=0)
+        except:
+            result = None
+        if not result or result.get('Result') != 'OK':
+            _LOGGER.debug('unable to select source')
+            return
+
+        self._source = source
+
+    async def _get_source(self):
+        self._source = None
+        self._source_list = None
+
+        action = self._device._action('MTVA', 'GetSourceList')
+        if not action:
+            _LOGGER.debug('Missing action MTVA/GetSourceList')
+            return
+
+        try:
+            result = await action.async_call()
+        except:
+            result = None
+        if not result or result.get('Result') != 'OK':
+            _LOGGER.debug('unable to get sources')
+            return
+
+        from collections import OrderedDict
+        from xml.dom.minidom import parseString
+        dom = parseString(result.get('SourceList'))
+        self._source = dom.getElementsByTagName('CurrentSourceType')[0] \
+                          .firstChild.nodeValue
+        self._source_list = OrderedDict()
+        for node in dom.getElementsByTagName('Source'):
+            con = node.getElementsByTagName('Connected')[0].firstChild\
```

Comment:
This is too much protocol specific information in this PR, this needs to be extracted into it's own PyPI library.